### PR TITLE
Add CloudEvents extension for deprecated events

### PR DESCRIPTION
**Pull request by vandewillysilva**
Introduced the `deprecation` attribute to indicate when an event type is deprecated, and added the `sunset` attribute to specify the date and time when the event will become unsupported. This extension provides clear guidelines and examples for implementing these attributes, aiming to improve lifecycle management and ensure better communication with consumers.

### CODE REVIEW

File: cloudevents/extensions/deprecation.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+- Examples:
+  - A boolean deprecation: `"deprecation": "true"`
+  - A date-based deprecation: `"deprecation": "Sun, 11 Aug 2024 23:59:59 GMT"`
+
+### sunset
+
+- Type: `String`
+- Description: Specifies the future date and time when the event type will
+  become unsupported, formatted according to
+  [RFC 8594](https://tools.ietf.org/html/rfc8594).
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:42]: While I can see the purpose of using RFC 8594 here, CE isn't tightly bound to HTTP, and as it already has a Timestamp format. I'd definitely prefer this to be a Timestamp-typed attribute, whatever we call it.

File: cloudevents/extensions/deprecation.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+needs to be in all CloudEvents, rather it needs to be included only when this
+extension is being used.
+
+## Attributes
+
+### deprecation
+
+- Type: `String`
+- Description: Indicates that the event type is deprecated. This can be a
+  boolean `true` or a date in [RFC 5322](https://tools.ietf.org/html/rfc5322)
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:28]: Rather than having a String that can be one of two formats, I'd prefer to have two attributes, each of which uses a single existing CE type - one `Boolean`, one `Timestamp`. (e.g. `deprecated` and `deprecatedfrom`)

File: cloudevents/extensions/deprecation.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  format indicating when the deprecation was announced.
+- Constraints
+  - REQUIRED
+  - If the value is a date, it MUST be formatted according to RFC 5322.
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:32]: Why introduce a different (and relatively hard to parse, culture-sensitive) format when we already use RFC3339 for the CloudEvent Timestamp format?

File: cloudevents/extensions/deprecation.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+- Examples:
+  - A boolean deprecation: `"deprecation": "true"`
+  - A date-based deprecation: `"deprecation": "Sun, 11 Aug 2024 23:59:59 GMT"`
+
+### sunset
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:37]: If we do end up with two attributes above, I'd prefer this to be something starting with "deprecated" or "deprecation" - perhaps "deprecationsunset" is best, but it keeps them all together.

File: cloudevents/extensions/deprecation.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+## Usage
+
+When this extension is used, producers MUST set the value of the `deprecation`
+attribute to `true` or the `timestamp` of when an event is being phased out but
+still supported. This gives consumers a heads-up that they SHOULD begin
+migrating to a new event type or version.
+
+The `sunset` attribute SHOULD specify the exact date and time when the event
+will no longer be supported. This is the final cutoff date after which the
+event will no longer function as expected.
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:58]: We should probably have some guidance around what different actors might do with an event that they see after this time. For example, if you're an "event forwarder" should you just drop such events?

File: cloudevents/extensions/deprecation.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+"MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT",
+"RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as
+described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+However, the scope of these key words is limited to when this extension is
+used. For example, an attribute being marked as "REQUIRED" does not mean it
+needs to be in all CloudEvents, rather it needs to be included only when this
+extension is being used.
+
+## Attributes
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:22]: I'd like some sort of attribute describing a migration path - "stop using this, start using X". We could make that a URI instead of text, for people to link to docs. Not sure of the details, but something would be good.

**Review by jskeet — COMMENTED** [review]
Definitely happy with the idea of being able to signal deprecation - definitely don't want to introduce more timestamp formats.

File: cloudevents/extensions/deprecation.md (hunk 4 above)
Comment (ttripp) [reply cloudevents/extensions/deprecation.md:37]: Although a different purpose, the `Deprecation` name is aligned with [this IETF draft standard](https://datatracker.ietf.org/doc/html/draft-ietf-httpapi-deprecation-header-04).

Comment (jskeet) [reply cloudevents/extensions/deprecation.md:37]: Again, that's HTTP-centric though. While I don't mind if attribute names happen to coincide with HTTP ones, I don't think that should be a driving factor.

File: cloudevents/extensions/deprecation.md (hunk 1 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md]: Changed type to `Timestamp`

File: cloudevents/extensions/deprecation.md (hunk 2 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md]: separated out the attributes:
- deprecation
- deprecationfrom

File: cloudevents/extensions/deprecation.md (hunk 3 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md]: fixed

File: cloudevents/extensions/deprecation.md (hunk 4 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md]: Renamed to deprecationsunset

File: cloudevents/extensions/deprecation.md (hunk 6 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md:21]: Added a new (optional) attribute: `deprecationmigration`

File: cloudevents/extensions/deprecation.md (hunk 5 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md]: Added a new paragraph in the `Usage` section. Please take a look at it.

File: cloudevents/extensions/deprecation.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+### deprecation
+
+- Type: `Boolean`
+- Description: Indicates whether the event type is deprecated.
+- Constraints
+  - REQUIRED
+  - The value MUST be a boolean (`true` or `false`).
+- Example: `"deprecation": "true"`
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:30]: I'm torn between the consistency of having a common prefix, and the oddity of it sounding like it's not Boolean (whereas "deprecated: true" sounds fine).

Thinking about it more, when would this ever actually have a value of false? That would be unusual at least...

File: cloudevents/extensions/deprecation.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+### deprecationfrom
+
+- Type: `Timestamp`
+- Description: Specifies the date and time when the event type was
+  officially marked as deprecated.
+- Constraints
+  - OPTIONAL
+  - If present, MUST adhere to the format specified in
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:39]: I suspect we don't need this, because it's inherent in the type. (And only the string representation needs to adhere to a format anyway - if this is formatted with a Protobuf formatter for example, it'll end up as a protobuf timestamp.)

Ditto deprecationsunset

File: cloudevents/extensions/deprecation.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+- Example: `"deprecationmigration": "https://example.com/migrate-to-new-evt"`
+
+## Usage
+
+When this extension is used, producers MUST set the value of the `deprecation`
+attribute to `true`. This gives consumers a heads-up that they SHOULD begin
+migrating to a new event or version.
+
+Consumers SHOULD use the `deprecation` and `deprecationfrom` attributes to
+monitor the lifecycle of events they rely on. If `deprecationsunset` has been
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:75]: Presumably consumers should consider switching to the recommended replacement as soon as they see that the event is deprecated, really? (You don't want to wait until the sunset.) Perhaps this should be more "consumers should make efforts to switch to the recommended replacement before the specified sunset"?

File: cloudevents/extensions/deprecation.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+reached, consumers SHOULD consider switching to the RECOMMENDED replacement, as
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:76]: I don't think RECOMMENDED needs to be in capitals here.

File: cloudevents/extensions/deprecation.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+  - If present, MUST adhere to the format specified in
+    [RFC 3339](https://tools.ietf.org/html/rfc3339)
+- Example: `"deprecationfrom": "2024-10-11T00:00:00Z"`
+
+### deprecationsunset
+
+- Type: `Timestamp`
+- Description: Specifies the future date and time when the event type will
+  become unsupported.
+- Constraints
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:48]: I wonder whether it's worth constraining this to say that the sunset for a type should never be observed to become earlier, but may be observed to become later. (An event producer may choose to support an event for longer than initially expected, but shouldn't shorten that support window.)

File: cloudevents/extensions/deprecation.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+  - REQUIRED
+  - The value MUST be a boolean (`true` or `false`).
+- Example: `"deprecation": "true"`
+
+### deprecationfrom
+
+- Type: `Timestamp`
+- Description: Specifies the date and time when the event type was
+  officially marked as deprecated.
+- Constraints
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:37]: Should we have a constraint around the stability of this? (It would be odd for it to vary over time.)

Should this always be in the (potentially very near) past or does it make sense to pre-announce deprecation?

**Review by jskeet — COMMENTED** [review]
This looks a *lot* better to me, thanks so much. Various additional questions...

**Comment by Fannon** [comment]
@vandewillysilva : Sorry, I didn't have time to look deeper into it, but just wanted to drop some links:

Here is a RFC for sunset as a header: https://datatracker.ietf.org/doc/html/rfc8594

At my company, SAP we have also standardized the same attributes as you propose here. In case you want to have a look, they're documented in the Open Resource Discovery specification, here: https://sap.github.io/open-resource-discovery/spec-v1/interfaces/document#event-resource_releasestatus

In particular, check `releaseStatus` (has deprecated as one value), `deprecationDate`, `sunsetDate`, `successors`. 

Also, we noticed that deprecating a resource does not necessarily have to imply a sunset. Of course you have to deprecate first, and sometimes you already plan the sunset when you do. But sometimes, you just created a successor resource and want just to indicate that the old resource is deprecated and the newer should be used.

File: cloudevents/extensions/deprecation.md (hunk 7 above)
Comment (Fannon) [reply cloudevents/extensions/deprecation.md:30]: Does this have to be REQUIRED? 

If you just have a `deprecated`: true and imply a default of false, there is no need to add a mandatory attribute.

File: cloudevents/extensions/deprecation.md (hunk 8 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md:39]: we have that in the recordedtime extension: https://github.com/cloudevents/spec/blob/main/cloudevents/extensions/recordedtime.md?plain=1#L45-L46

I was just following the same convention

File: cloudevents/extensions/deprecation.md (hunk 10 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md:76]: there is a lint script that complains about that. I'll reword it to `suggested`.

File: cloudevents/extensions/deprecation.md (hunk 8 above)
Comment (jskeet) [reply cloudevents/extensions/deprecation.md:39]: I'll happily create a PR to get rid of it from there :)

File: cloudevents/extensions/deprecation.md (hunk 12 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md:38]: added a new constraint.

File: cloudevents/extensions/deprecation.md (hunk 11 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md:52]: added a constraint.

File: cloudevents/extensions/deprecation.md (hunk 9 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md]: Rephrased this paragraph. Please take a look at it.

File: cloudevents/extensions/deprecation.md (hunk 7 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md]: renamed to deprecated

File: cloudevents/extensions/deprecation.md (hunk 8 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md]: Remove reference to RFC 3339

**Comment by vandewillysilva** [comment]
> @vandewillysilva : Sorry, I didn't have time to look deeper into it, but just wanted to drop some links:
> 
> Here is a RFC for sunset as a header: https://datatracker.ietf.org/doc/html/rfc8594
> 
> At my company, SAP we have also standardized the same attributes as you propose here. In case you want to have a look, they're documented in the Open Resource Discovery specification, here: https://sap.github.io/open-resource-discovery/spec-v1/interfaces/document#event-resource_releasestatus
> 
> In particular, check `releaseStatus` (has deprecated as one value), `deprecationDate`, `sunsetDate`, `successors`.
> 
> Also, we noticed that deprecating a resource does not necessarily have to imply a sunset. Of course you have to deprecate first, and sometimes you already plan the sunset when you do. But sometimes, you just created a successor resource and want just to indicate that the old resource is deprecated and the newer should be used.

@Fannon thanks for the links and insights. It's great to see that we're aligned on similar attributes. :)

File: cloudevents/extensions/deprecation.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+## Attributes
+
+### deprecated
+
+- Type: `Boolean`
+- Description: Indicates whether the event type is deprecated. Defaults to
+  `false` if not specified.
+- Constraints
+  - OPTIONAL
+  - The value MUST be a boolean (`true` or `false`).
```
Comment (jskeet) [inline cloudevents/extensions/deprecation.md:30]: Just like the timestamp part, I don't think we need to say this - it's implicit in the type being Boolean.

**Review by jskeet — APPROVED** [review]
One more nit - but basically I'm otherwise happy :)

File: cloudevents/extensions/deprecation.md (hunk 13 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md]: agreed. Removed the constraint

**Comment by santhoshhpe** [comment]
A must have feature

**Review by santhoshhpe — APPROVED** [review]

**Comment by vandewillysilva** [comment]
@duglin Do you think it’s a good idea to proceed with this deprecation extension proposal? Let me know if there’s any other work needed or if you have any additional feedback.

File: cloudevents/extensions/deprecation.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+## Attributes
+
+### deprecated
+
+- Type: `Boolean`
+- Description: Indicates whether the event type is deprecated. Defaults to
+  `false` if not specified.
+- Constraints
+  - OPTIONAL
+- Example: `"deprecated": "true"`
```
Comment (duglin) [inline cloudevents/extensions/deprecation.md:30]: `"deprecated": true`
no quotes

File: cloudevents/extensions/deprecation.md [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+### deprecationfrom
+
+- Type: `Timestamp`
+- Description: Specifies the date and time when the event type was
+  officially marked as deprecated.
+- Constraints
+  - OPTIONAL
+  - The `deprecationfrom` timestamp SHOULD remain stable once set and SHOULD
```
Comment (duglin) [inline cloudevents/extensions/deprecation.md:39]: Just curious, aside from this being a possible "FYI" bit of info, do people use this timestamp for some purpose in real life? Doesn't seem like it would influence people's action.... the "sunset" one is the important one.

File: cloudevents/extensions/deprecation.md [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+
+### deprecationmigration
+
+- Type: `URI`
+- Description: Provides a link to documentation or resources that describe
+the migration path from the deprecated event to an alternative. This helps
+consumers transition away from the deprecated event.
+- Constraints
+  - OPTIONAL
+  - The URI SHOULD point to valid and accessible resources that help the
```
Comment (duglin) [inline cloudevents/extensions/deprecation.md:64]: s/help/helps/

File: cloudevents/extensions/deprecation.md (hunk 15 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md:39]: the `deprecationsunset` dictates the hard deadline, but `deprecationfrom` can prompt consumers to start planning their migration earlier.
IMO, it adds valuable context and encourages proactive behavior from consumers.

File: cloudevents/extensions/deprecation.md (hunk 14 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md]: fixed

File: cloudevents/extensions/deprecation.md (hunk 16 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md]: fixed

File: cloudevents/extensions/deprecation.md [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+However, the scope of these key words is limited to when this extension is
+used. For example, an attribute being marked as "REQUIRED" does not mean it
+needs to be in all CloudEvents, rather it needs to be included only when this
+extension is being used.
+
+## Attributes
+
+### deprecated
```
Comment (duglin) [inline cloudevents/extensions/deprecation.md:23]: Since we basically say that even REQUIRED fields are really optional if the extension isn't used, I'm wondering if this one should be REQUIRED. It would be kind of weird to have this one "false" but any of the other attributes present, right?

File: cloudevents/extensions/deprecation.md (hunk 15 above)
Comment (duglin) [reply cloudevents/extensions/deprecation.md:39]: In theory, but I would think the presence of any of these would indicate that the event is deprecated regardless of when the deprecation period stated, right?

**Comment by duglin** [comment]
> @duglin Do you think it’s a good idea to proceed with this deprecation extension proposal? Let me know if there’s any other work needed or if you have any additional feedback.

Yup - it's on the agenda for tomorrow's call. Will you be on the call? Not a requirement but it would be nice :-)
https://docs.google.com/document/d/1OVF68rpuPK5shIHILK9JOqlZBbfe91RNzQ7u_P7YCDE/edit#heading=h.6xidlvhv1goi

File: cloudevents/extensions/deprecation.md (hunk 15 above)
Comment (vandewillysilva) [reply cloudevents/extensions/deprecation.md:39]: that's correct.
`deprecationfrom` adds an extra layer of clarity. This can be useful for teams who need to plan their migrations earlier rather than waiting until the sunset approaches.

**Comment by vandewillysilva** [comment]
PR is failing due to broken links in avro spec. This PR #1309 fixes the issue.

**Comment by vandewillysilva** [comment]
@duglin I went over the recording and made the changes related to the deprecated field. Please take a look at it.

File: cloudevents/extensions/deprecation.md [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+
+- Type: `Boolean`
+- Description: Indicates whether the event type is deprecated.
+- Constraints
+  - MUST be true
```
Comment (duglin) [inline cloudevents/extensions/deprecation.md:28]: s/true/`true`/g

**Comment by duglin** [comment]
@vandewillysilva just one minor tweak and then we can merge

**Comment by vandewillysilva** [comment]
> @vandewillysilva just one minor tweak and then we can merge

Good catch! Fixed.

**Comment by duglin** [comment]
Approved on the 8/22 call.
Thanks @vandewillysilva
